### ISSUE TITLE
temporarily bypass the call rate limit in public query?

### ISSUE THREAD

**Issue by guanqun** [hugues31/coinnect#36]
In some cases, I'd like to call these in a burst mode, I don't want to be blocked by the API internally...

**Comment by ainestal** [comment]
If you do a burst and after it you fall back to the standard ratio, there are chances you will be banned. The only way I see to do it is by allowing a burst and then allowing some time to recover from it. This doesn't look simple because it depends on the length and amount of requests of the burst.

It is also dependant on the different exchanges, each one having a different way of controlling the amount of requests issued.

That being said, I understand the convenience of having this feature, I already had a similar need.

I still haven't a strong opinion on this, at this point I would like to hear other opinions and options.

**Comment by hugues31** [comment]
Hello! We can always implement a method on ExchangeApi like `set_block_or_continue(bool)` which bypass the `block_or_continue` function. The `block_or_continue` function itself can be improved by logging all queries and make the proper sleep duration following the rules imposed by each exchanges.

**Comment by guanqun** [comment]
The caller of this API will make sure to wait enough time before sending more requests to it. Therefore we won't be banned, yeah, if the caller fails to do so, then definitely we'll be banned...

If you agree adding a `set_block_or_continue(bool)` is fine, I'll help add this.

**Comment by ainestal** [comment]
I like the idea of having a way of controlling the time from the perspective of the user. In that case, as a user, I would like to have full control over the time, so no need to keep track of the requests from the library. That also would vastly simplify the library.

**Comment by hugues31** [comment]
I'm ok to add `set_block_or_continue(bool)` function. This feature is helpful & easy to implement.
@guanqun if you submit a PR for this, I'll be happy to merge your work.

Furthermore, I suggest to improve the `block_or_continue` function to respect more the exchanges rules but we could open a new issue for this as there is no simple solution. (for example, Kraken has an API call rate limit depending on user account and query type so this is not trivial)

**Comment by ainestal** [comment]
I've been working on this for a bit and I found something really interesting: time doesn't get the correct nanoseconds. I created a ticket but the library is deprecated so I don't know when or if they will solve it. You can see the issue here:
https://github.com/rust-lang-deprecated/time/issues/154

I don't trust the time library anymore, we should use [chrono](https://crates.io/crates/chrono) instead.

**Comment by hugues31** [comment]
By no mean I'm an expert but I don't think standard CPU could reach the nanosecond precision, doesn't the same thing happen with chrono ? If not, we could open an issue to migrate Coinnect to chrono :)

**Comment by ainestal** [comment]
I did the same test with chrono and there is a difference of some nanoseconds, it works. I'm going to open the issue like you suggested.

**Comment by ainestal** [comment]
#45 Implements the bypass of `block_or_continue` with `set_burst(bool)`

**Comment by hugues31** [comment]
PR merged. I close.
